src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1015,3 +1069,919 @@ def upgrade_agents(cmd, client, resource_group_name, cluster_name, kube_config=N
+
+    subscription_id = get_subscription_id(cmd.cli_ctx)
+
+    if token is not None:
+        auth_method = 'Token'
+        telemetry.add_extension_event('connectedk8s', {'Context.Default.AzureCLI.IsAADEnabled': False})
+    else:
+        auth_method = 'AAD'
+        telemetry.add_extension_event('connectedk8s', {'Context.Default.AzureCLI.IsAADEnabled': True})
+
+    # Fetching hybrid connection details from Userrp
+    try:
+        response = client.list_cluster_user_credentials(resource_group_name, cluster_name, auth_method, True)
+    except Exception as e:
+        telemetry.set_exception(exception=e, fault_type=consts.Get_Credentials_Failed_Fault_Type,
+                                summary='Unable to list cluster user credentials')
+        raise CLIError("Failed to get credentials." + str(e))
+
+    # Starting the client proxy process, if this is the first time that this function is invoked
+    if flag == 0:
+        if debug_mode:
+            clientproxy_process = Thread(target=call, args=(args,))
+        else:
+            clientproxy_process = Thread(target=call, args=(args,), kwargs={'stderr': DEVNULL, 'stdout': DEVNULL})
+
+        try:
+            clientproxy_process.start()
+            print(f'Proxy is listening on port {api_server_port}')
+        except Exception as e:
+            telemetry.set_exception(exception=e, fault_type=consts.Run_Clientproxy_Fault_Type,
+                                    summary='Unable to run client proxy executable')
+            raise CLIError("Failed to start proxy process." + str(e))
+
+        if user_type == 'user':
+            identity_data = {}
+            identity_data['refreshToken'] = creds
+            identity_uri = f'https://localhost:{api_server_port}/identity/rt'
+
+            # Needed to prevent skip tls warning from printing to the console
+            original_stderr = sys.stderr
+            f = open(os.devnull, 'w')
+            sys.stderr = f
+
+            make_api_call_with_retries(identity_uri, identity_data, False, consts.Post_RefreshToken_Fault_Type,
+                                       'Unable to post refresh token details to clientproxy',
+                                       "Failed to pass refresh token details to proxy.")
+            sys.stderr = original_stderr
+
+    data = prepare_clientproxy_data(response)
+    expiry = data['hybridConnectionConfig']['expirationTime']
+
+    if token is not None:
+        data['kubeconfigs'][0]['value'] = insert_token_in_kubeconfig(data, token)
+
+    # Starting a timer to refresh the credentials, 5 mins before expiry
+    fun_args = [cmd, client, resource_group_name, cluster_name, 1, args, client_proxy_port, api_server_port, operating_system, creds, user_type, debug_mode, token, path, overwrite_existing, context_name]
+    refresh_thread = Timer(expiry - time.time() - 300, client_side_proxy, args=fun_args)
+    refresh_thread.setDaemon(True)
+    try:
+        refresh_thread.start()
```

Comment:
I see the in the timer we are recursively calling the same function and we are starting a new timer and running it on a new thread?. So how will the previous thread end ?.

Reply:
The previous thread will end on its own after it has refreshed the creds, because nothing else is left to be executed for that thread. I have verified this. Active count of threads is at max 4.